// grunt-ejs-static/lib/ejs_static.js

var EJS_Static = function() {

  // private functions
  // 
  var data_functions = require('./data'),
      template_functions = require('./template'),
      layout_functions = require('./layout'),
      render_functions = require('./render'),
      utility_functions = require('./utils');

  // 
  // end private functions

  // public functions
  // 
  return {

    // data
    data_validator: data_functions.data_validator,
    parse_data_file: data_functions.parse_data_file,
    prepare_data: data_functions.prepare_data,
    data_type_router: data_functions.data_type_router,

    // template
    prepare_template: template_functions.prepare_template,

    // layout
    path_to_layout_validator: layout_functions.path_to_layout_validator,
    path_to_layout_generator: layout_functions.path_to_layout_generator,
    search_layouts_dir: layout_functions.search_layouts_dir,
    prepare_layout: layout_functions.prepare_layout,

    // render
    create_files_router: render_functions.create_files_router,
    create_index_file: render_functions.create_index_file,
    create_category_file: render_functions.create_category_file,
    create_root_level_file: render_functions.create_root_level_file,
    render_template: render_functions.render_template,

    // utils
    underscores_to_dashes: utility_functions.underscores_to_dashes
  
  }
  // 
  // end public functions

}

module.exports = exports = new EJS_Static();
