// grunt-ejs-static/lib/validate_data.js

module.exports = exports = {

  // validate data before rendering
  validate_data: function(ejs_static) {

    var grunt = ejs_static.grunt;

    data_validator(ejs_static);

    // PRIVATE FUNCTIONS
    // 

    function data_validator(ejs_static) {

      // get data specified in ejs_static options
      if (typeof ejs_static.options.path_to_data === 'string') {

        // get the data
        var data = grunt.file.readJSON(ejs_static.options.path_to_data);

        // parse the data
        parse_data_files(data);

        grunt.log.debug('Data successfully imported');
        
      } else {

        grunt.fail.warn('The path_to_data option is required, please specify in Gruntfile');
        return false;

      }
      // end get data specified in ejs_static options

    }

    // iterate through items in data file
    // http://stackoverflow.com/questions/7440001/iterate-over-object-keys-in-node-js
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys
    function parse_data_files(data_object) {

      Object.keys(data_object).forEach(function(key) {

        path_to_layout_validator(key, data_object);

      });

    }
    // end iterate through items in data file

    // search the layouts dir for a dir or file that matches key
    function search_layouts_dir(key) {

      var search_pattern1 = ejs_static.path.join(ejs_static.options.path_to_layouts, '**', key + '.ejs');
      grunt.log.debug('search pattern1 is ' + search_pattern1);

      var search_pattern2 = ejs_static.path.join(ejs_static.options.path_to_layouts, '**', key, 'index' + ejs_static.options.file_extension);
      grunt.log.debug('search pattern2 is ' + search_pattern2);

      var search_pattern3 = ejs_static.path.join(ejs_static.options.path_to_layouts, '**', key, 'index.ejs');
      grunt.log.debug('search pattern3 is ' + search_pattern3);

      var path_to_layout = grunt.file.expand(search_pattern1, search_pattern2, search_pattern3);

      // if exactly one path to layout is returned, return that path
      if (path_to_layout.length === 1) {

        grunt.log.debug('PATH TO LAYOUT is ' + path_to_layout);
        return path_to_layout;

      // if more than one path to layout is returned, throw error
      } else if (path_to_layout.length > 1) {

        var multiple_layout_paths = "";

        Object.keys(path_to_layout).forEach(function(key) {
          multiple_layout_paths += path_to_layout[key] + '\n';
        });

        grunt.fail.warn('More than 1 path to layout file was not found for key: ' + key + 
                        '\nThe paths found were: \n' + multiple_layout_paths + 
                        '\nResolve in the project dir structure and/or data file\n');

      // if no path to layout is returned, throw error
      } else {
        grunt.fail.warn('path to layout file was not found for key: ' + key);
      }

    }
    // end search the layouts dir for a dir or file that matches key

    // generate path to layout
    function path_to_layout_generator(key, data_object) {

      // if path_to_layout exists
      if (typeof data_object[key].path_to_layout === 'string') {

        return data_object[key].path_to_layout;

      // if a layouts dir is specified
      } else if (typeof ejs_static.options.path_to_layouts === 'string') {

        return search_layouts_dir(key, ejs_static);

      } else {
        grunt.fail.warn('A value for path_to_layout is required, please specify in data file for key: ' + key);
      }      

    }
    // end generate path to layout

    // test for path_to_layout value for key in data.json
    function path_to_layout_validator(key, data_object) {

      var path_to_layout = path_to_layout_generator(key, data_object);

      if (typeof path_to_layout !== 'undefined') {
        // data and layout have been validated, so start create file process
        ejs_static.create_file(key, data_object, path_to_layout, ejs_static);
      } else {
        grunt.log.debug('file NOT CREATED for key: ' + key);
      }

    }
    // end test for path_to_layout value for key in data.json

    // 
    // END PRIVATE FUNCTIONS

  }
  // end validate data before rendering

};
