// grunt-ejs-static/tasks/ejs_static.js
/*
 * grunt-ejs-static
 * https://github.com/shaekuronen/grunt-ejs-static
 *
 * Copyright (c) 2013 Shae Kuronen
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  var ejs = require('ejs');
  var path = require('path');

  // Please see the Grunt documentation for more information regarding task
  // creation: http://gruntjs.com/creating-tasks

  grunt.registerMultiTask('ejs_static', 'Render EJS templates as static HTML.', function() {

    // Merge task-specific and/or target-specific options with these defaults.
    var options = this.options({
      punctuation: '.',
      separator: ', '
    });

    // DATA PROCESSING
    var data = process_data(options.data);

    // combine data in options.data into a single data object
    function process_data(unprocessed_data) {

      var the_data = {};

      for (var key in unprocessed_data) {

        // make sure property is not from the prototype
        if (unprocessed_data.hasOwnProperty(key)) {

          var this_data = grunt.file.readJSON(unprocessed_data[key]);

          the_data[key] = this_data;

        } else {

          grunt.log.debug(key + ' failed hasOwnProperty');

        }

      }

      return the_data;

    }

    grunt.log.writeln('');

    grunt.log.writeln('The processed data is ' + JSON.stringify(data) );

    grunt.log.writeln('');

    grunt.log.writeln(data.global.google_analytics);
    // END DATA PROCESSING

    // get the page names from routes.json data
    function get_page_names(data_object) {

      var the_pages = {};

      for (var key in data_object.routes) {

        if ( data_object.routes[key].hasOwnProperty('category') ) {
          
          grunt.log.writeln('the page has a category');

        } else {

          grunt.log.writeln('the page has no category');

        }

        // convert any underscores into dashes
        var the_page_name = key.replace(/_/g,"-");

        grunt.log.writeln(the_page_name);
      } 

    }

    var pages = get_page_names(data);






    // get the parent directory of layout manager 
    function get_parent_directory(path_to_file) {

        var tokens = [];

        tokens = path_to_file.split('/');

        // remove the index.html file from array
        var the_file = tokens.splice(-1, 2);

        var the_parent_directory = tokens.pop();

        return the_parent_directory;

    }
    // end get the parent directory of layout manager

    // get the path to the rendered template
    function get_path_to_rendered_template(path_to_file, path_to_layout, destination) {

      // remove src (options.src) from beginning of path
      var path_to_rendered_template = path_to_file.replace(path_to_layout, "");

      // add destination (options.dest) to beginning of path
      path_to_rendered_template = destination + path_to_rendered_template;

      return path_to_rendered_template;     
      
    }
    // end get the path to the rendered template

  });

};
